File: file-translation-demo/nodejs-js/file-translate-download.js
/**
* Demo program of download translated file.
* Command line: node ./file-translation-download.js translateItemId1,translateItemId2,translateItemId3....
* Required npm libs: superagent, crypto, adm-zip
*/
const superagent = require('superagent');
const crypto = require('crypto');
const fs = require('fs');
const AdmZip = require('adm-zip');

const { serverConfig } = require('./config');
const { authConfig } = require('./config');

const signatureHMACAlgo = 'sha256';
const signatureHMACEncoding = 'hex';
/**
 * Generates a request signature.
 *
 * @param {string} path Path.
 * @param {string} secretKey Secret key.
 * @param {string} nonce Nonce.
 *
 * @returns {string} The request signature.
 */
const generateSignature = (path, secretKey, nonce) => {
  const hmac = crypto.createHmac(signatureHMACAlgo, secretKey);
  hmac.update(nonce);
  hmac.update(path);
  return hmac.digest(signatureHMACEncoding);
};

/**
* @param {object} serverConfig Server configurations.
* @param {string} serverConfig.protocol Server protocol.
* @param {string} serverConfig.hostname Server hostname.
* @param {number} serverConfig.port Server listening port.
* @param {object} authConfig Authentication configurations.
* @param {string} authConfig.accessKey Access key.
* @param {string} authConfig.secretKey Secret key.
* @param {string} authConfig.nonce Nonce.
* @param {string} translateItemIds Translate item ids to be downloaded, splitted by ','
*
* @returns {Promise<string>} Server response.
*
* @throws {Error} When unable to complete the request.
*/
const sendRequest = (serverConfig, authConfig, translateItemIds) => {
  const itemIds = translateItemIds.split(',')
  const url = `/api/v1/downloads?ids=${JSON.stringify(itemIds)}`;
  const nonce = new Date().getTime().toString();
  const signature = generateSignature(url, authConfig.secretKey, nonce);

  superagent.get(`${serverConfig.protocol}//${serverConfig.hostname}${url}`)
    .set({
      accessKey: authConfig.accessKey,
      signature,
      nonce,
    }).end(function (req, resp) {
      if (resp.status === 200) {
        fs.createWriteStream('./output.zip').write(resp.body, (error) => {
          if (error) {
            console.error(error);
          } else {
            const zip = new AdmZip('./output.zip');
            zip.extractAllTo('./', true);
          }
        });
      }
    });
};

const main = async () => {
  const translateItemIds = process.argv[2];
  if (!translateItemIds) {
    console.log("Error. please input translationItemId.");
    return;
  };
  try {
    await sendRequest(
      serverConfig,
      authConfig,
      translateItemIds
    );
  } catch (error) {
    console.error(error);
  }
};

main();

File: file-translation-demo/python/file-translate-download.py
import time
import os
import json
import requests
import zipfile
import auth_utils
import config

def send_request(server_config, access_key, secret_key, translate_item_ids):
    item_ids = translate_item_ids.split(",")
    url = f"/api/v1/downloads?ids={json.dumps(item_ids)}"
    nonce = str(int(time.time()))
    signature = auth_utils.generate_signature(url, secret_key, nonce)

    headers = {
        "accessKey": access_key,
        "signature": signature,
        "nonce": nonce,
    }

    response = requests.get(
        f"{server_config['protocol']}//{server_config['hostname']}{url}",
        headers=headers
    )
    if response.status_code == 200:
        with open("./output.zip", "wb") as f:
            f.write(response.content)
        with zipfile.ZipFile("./output.zip") as zip_file:
            zip_file.extractall("./")
        os.remove("./output.zip")

def main():
    # response.json()['data']['items'][0]['translateItemId']
    # or
    # response.json()['data']['items'][0]['translateItemId'],response.json()['data']['items'][2]['translateItemId'],...
    translate_item_ids = "your translateItemId"
    try:
        send_request(
            config.server_config,
            config.auth_config['access_key'],
            config.auth_config['secret_key'],
            translate_item_ids
        )
    except Exception as error:
        print(error)

if __name__ == "__main__":
    main()
